beets/config_default.yaml

```diff
@@ -118,6 +118,10 @@ musicbrainz:
     searchlimit: 5
     extra_tags: []
     genres: no
+    url_rels:
+        discogs: no
+        spotify: no
```

Comment:
I think we should have an `overwrite` flag as well. If a user has already matched using one of the other trackers, say `spotify`, we should not overwrite that information unless the user explicitly asks for that.

Reply:
it's in the works ;-)

Reply:
Alright, I'm not really sure what I'm doing here....does this sound like the right place at all to implement an overwrite flag: https://github.com/beetbox/beets/pull/4474/commits/8f864b44f2e70c79fb7b9fc4a70bbfb5b0245547

Reply:
To be honest, I probably should bother @sampsyo here: With above's commit, why do I get an item that for example has an empty discos_albumid field even though beet info -l shows that it actually is populated:

Debug print of the item inside the apply_metadata function.
```
discogs_albumid=0
```

but beet info is showing that it's populated in the db:

```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info makoto golden -l | grep discogs_album
     discogs_albumid: 389624
     discogs_albumid: 389624
(beets)    ~/git/beets   mb_fetch_discogs_url ●  
```

Reply:
The most likely culprit is that this information is present only at the album level, not the track level (or vice versa). Is there a chance that the field is only applied to one or the other?

Reply:
Hi! Thanks so much for the quick response and putting my in the right direction! Funnily enough yesterday evening I did another poking around debugging session and found out exactly that! Yes indeed this field is only set on the album level and not on the items!

Reply:
Ok I answered to quickly. Again I confused the album_info with the mapping. I still don't know how to get the "current discogs_albumid set in the beets database already" in this place in the code. I think it is just not possible here and I need to look somewhere else.... Looking into the database directly I find the information saved in the album table but as well on each track in the items table!

Spelling it out loudly:

album_info is the new information we just fetched and set in autottag.mb.album_info() -> We only have new information in here and can't compare with old information!

mapping seems to be, at least to my (probably wrong) understanding, the current information we have in the database already in the form of TrackInfo objects. It does not contain the discogs_albumid.

Reply:
After around seven hundred debug messages all around beets code I found out that:

When reimporting via the file path:
```
beet -v import ~/Sync/beets_test_music/Good\ Looking\ -\ Makoto\ Featuring\ MC\ Conrad
```
the discogs_albumid attribute is not set on the TrackInfo object because on my test file it is not existing as a file tag but  _only_ set in the beets db!

When reimporting from file the construction of an Item object seems to work a little different. Trying to understand what exactly is happening here at the moment.... https://github.com/beetbox/beets/blob/master/beets/library.py#L602-L604

On the other hand, when reimporting via a library query:

```
beet -v import -L makoto golden
```

my code submitted in https://github.com/beetbox/beets/commit/8f864b44f2e70c79fb7b9fc4a70bbfb5b0245547 works perfectly fine: The Trackinfo object _does_ have the discogs_albumid property set! (Even though it is _only_ available as a database field and _not_ available from the media file's tag)

So actually what I need to find out currently, is:
- Why is discogs_albumid not written to the media file when it is populated with this PR's new feature?
- Is it expected behaviour that an Item object when reimported via the file path, does not contain attributes that are only set in the db, but are not available from the file's tags.

Reply:
> * Why is `discogs_albumid` not written to the media file when it is populated with this PR's new feature?

The answer is obvious, I now recall how writing to media files works in beets: discogs_albumid is never written to disk, since the Discogs plugin does not add the MediaField. This could be made possible with something like that in the plugin's initialization (but that's probably something for another PR, just saving the thougth):

```
        self.add_media_field("discogs_albumid", MediaField(
            MP3DescStorageStyle(u'discogs_albumid'),
            StorageStyle(u'discogs_albumid')
        )
        # and further storage styles for MP4 and ASF
```

Also I realize that none of the metadatasource plugins do add media_fields, thus with the current state of beets we never get file tags `spotify_album_id` `deezer_album_id`, `beetport_album_id`, and so on. Also there currently is no way of configuring that these fields should be written to disk. Correct my if I'm wrong!

Reply:
Ok, I'm about to give up. I don't understand why this is not working and it's getting pretty frustrating by now to not being able to achieve the simple task of storing something in a database. There must be something obvious I'm not seeing.

So please @sampsyo, @arogl , @arsaboo, anyone, ... can you have a look at this and tell me if by any means this could work: https://github.com/beetbox/beets/pull/4474/commits/7206083d6864d6f1109ec2c679d93151d7fc9590

The Initial import of the id fields via MusicBrainz works fine but as soon as I try to reimport with the external_ids_overwrite flag set to true, I get into trouble. It just doesn't seem to be possible to update the values in the database. I thought I had it, thought it goes wrong in this place because it overwrites the new values with values from the current items in the database but it doesn't do the trick.

So for example let's take this 2-track single:
```
$ beet -v import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/
...
...
$ beet info -l robert hood alarm | grep spotify
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV

$ beet -v edit robert hood alarm
# change one of the ID's to something else

$beet info -l robert hood alarm | grep spotif
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD

$ beet -v import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/ 
...
Reimported album: added 1666681278.518687, flexible attributes [] from album 31 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag
Reimported item added 1666681278.5210571 from item 75 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/01 - Robert Hood - Eleven - Gm,Abm.m4a
Reimported item flexible attributes ['spotify_album_id', 'track_alt', 'data_source', 'track_alt', 'spotify_album_id', 'data_source'] from item 75 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/01 - Robert Hood - Eleven - Gm,Abm.m4a
Sending event: database_change
Reimported item added 1666681278.524402 from item 76 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/02 - Robert Hood - Alarm - Am.m4a
Reimported item flexible attributes ['spotify_album_id', 'track_alt', 'data_source', 'track_alt', 'data_source', 'spotify_album_id'] from item 76 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/02 - Robert Hood - Alarm - Am.m4a

...

$ beet info -l robert hood alarm | grep spotif                                                        
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
```

Reply:
Can you try breaking up this part and adding debug logs to see if we have the right flow?

https://github.com/beetbox/beets/blob/7206083d6864d6f1109ec2c679d93151d7fc9590/beets/autotag/__init__.py#L195-L197

Specifically, create two if conditions for `if not item.get('spotify_album_id') or ext_id_force:`. Since the first import worked, I am assuming that the `not item.get('spotify_album_id')` is working and the tag writing is working fine.

Reply:
Hi @arsaboo, thanks so much for helping with debugging! I checked this condition a 100 times already. It is definitely working as expected and I'm pretty sure the problem lies somewhere else! Anyway, as suggested I torned apart the condition into two branches and put some debug print messages. This is the commit: 7f3a914050675244312fe1c98e4aa49bd15ad061, and this is the output

```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet rm robert hood alarm 
Robert Hood - Eleven / Alarm - Eleven
Robert Hood - Eleven / Alarm - Alarm
Really remove 2 items from the library?? (Yes/no/select) y

(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/  

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/ 

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import -L robert hood alarm                                                                 

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  

```

Reply:
Ok let's talk about problem number 2: Here first without any new debug messages, just to face the problem:

```
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info robert hood alarm | grep spotify                  
(beets)      ~/git/beets   mb_fetch_discogs_url  
(beets)      ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  beet edit robert hood alarm               
Robert Hood - Eleven / Alarm - Eleven
  spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV -> ABCD
continue [E]diting, Apply, Cancel? a
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm             

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  

```

Reply:
After a lot of debugging I found out that the problem might be in 'importer.ImportTask.reimport_metadata`, thus let's have a look at this debug commit 7206083d6864d6f1109ec2c679d93151d7fc9590 and it's output:


```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info robert hood alarm | grep spotify   
(beets)      ~/git/beets   mb_fetch_discogs_url ●  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import -L robert hood alarm             

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
We are in importer.ImportTask.reimport_metadata:
dup_item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: ABCD
We've now set item.spotify_album_id to: ABCD
We've also added spotify_album_id to item._dirty: {'discogs_albumid', 'added', 'spotify_album_id'}
dup_item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
We've now set item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'discogs_albumid', 'added', 'spotify_album_id'}
(beets)    ~/git/beets   mb_fetch_discogs_url ●  
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  

```

Reply:
@arsaboo I finally realized what my error was. I had to change dup_item instead of item in my last commit. This fixes it: 0f16e2a7483159c1380a1ba47d03c5e1886aac10

Here's the output:

```
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify                             
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm                                          

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
We are in importer.ImportTask.reimport_metadata:
item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
We've now set dup_item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'added', 'spotify_album_id', 'discogs_albumid'}
item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: ABCD
We've now set dup_item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'added', 'spotify_album_id', 'discogs_albumid'}
(beets)    ~/git/beets   mb_fetch_discogs_url  
```

Reply:
That is some amazing work. Kudos for not giving up and continuing to address the issue. 

Hopefully, the end is in sight. I am so excited to see this feature.

Reply:
@JOJ0 I also noticed that there were 2 spotify_album_id in the output. Is this normal or should one of them be a track_id like MB and discogs?

``` sh
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify                             
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm                                          
```

Reply:
@arogl I'm not sure if I understand the concern. You migth have missed that I'm testing with 2 media files and I'm changing one of them to ABCD. The other one I didnt touch. But you might be talking about something else

Reply:
@JOJ0 

All OK, I thought it was only 1 file with a duplicated tag.

I looked at the link and see it is a 2-track album.

Well done on the debugging